src/System.Net.Security/src/System/Net/SecureProtocols/_SslState.cs

```diff
@@ -0,0 +1,1745 @@
+             */
+
+            int version = -1;
+
+            GlobalLog.Assert((bytes != null && bytes.Length > 0), "SslState::DetectFraming()|Header buffer is not allocated will boom shortly.");
+
+            // If the first byte is SSL3 HandShake, then check if we have a SSLv3 Type3 client hello.
+            if (bytes[0] == (byte)FrameType.Handshake || bytes[0] == (byte)FrameType.AppData
+                || bytes[0] == (byte)FrameType.Alert)
+            {
+                if (length < 3)
+                {
+                    return Framing.Invalid;
+                }
+#if TRAVE
+                if (bytes[1] != 3)
+                    GlobalLog.Print("WARNING: SslState::DetectFraming() SSL protocol is > 3, trying SSL3 framing in retail = " + bytes[1].ToString("x", NumberFormatInfo.InvariantInfo));
+#endif
+
+                version = (bytes[1] << 8) | bytes[2];
+                if (version < 0x300 || version >= 0x500)
+                {
+                    return Framing.Invalid;
+                }
+
+                //
+                // This is an SSL3 Framing
+                //
+                return Framing.SinceSSL3;
+            }
+#if TRAVE
+            if ((bytes[0] & 0x80) == 0)
+                // We have a three-byte header format
+                GlobalLog.Print("WARNING: SslState::DetectFraming() SSL v <=2 HELLO has no high bit set for 3 bytes header, we are broken, received byte = " + bytes[0].ToString("x", NumberFormatInfo.InvariantInfo));
+#endif
+
+            if (length < 3)
+            {
+                return Framing.Invalid;
+            }
+            if (bytes[2] > 8)
+                return Framing.Invalid;
+
+            if (bytes[2] == 0x1)  // SSL_MT_CLIENT_HELLO
+            {
+                if (length >= 5)
+                    version = (bytes[3] << 8) | bytes[4];
+            }
+            else if (bytes[2] == 0x4) // SSL_MT_SERVER_HELLO
+            {
+                if (length >= 7)
+                    version = (bytes[5] << 8) | bytes[6];
+            }
+
+            if (version != -1)
+            {
+                // If this is the first packet, the client may start with an SSL2 packet
+                // but stating that the version is 3.x, so check the full range.
+                // For the subsequent packets we assume that an SSL2 packet should have a 2.x version.
+                if (_Framing == Framing.None)
+                {
+                    if (version != 0x0002 && (version < 0x200 || version >= 0x500))
+                        return Framing.Invalid;
+                }
+                else
+                {
+                    if (version != 0x0002)
+                        return Framing.Invalid;
+                }
+            }
+
+            // When server has replied the framing is already fixed depending on the prior client packet
+            if (!Context.IsServer || _Framing == Framing.Unified)
+            {
+                return Framing.BeforeSSL3;
```

Comment:
Behavior on Ubuntu is similar. SSLv2 is disabled but libssl can be rebuilt with explicit SSLv2 support